### PR TITLE
1.1.1k fips jra disable algos

### PR DESCRIPTION
**Pull request by jallisonciq**
Disable all FIPS-unapproved algorithms.

### CODE REVIEW

**Comment by jason-rodri** [comment]
list of disabled algos:
camellia
CAST
DES
DES3
DSA
IDEA
MD2
MD4
MD5
POLY1305
RC2
RC4
RC5
RIPEMD
WHIRLPOOL
siphash
Seed
SHA1
AES-OCB
ECDH/ECDSA  P192, K163, B163

File: crypto/rsa/rsa_oaep.c [hunk 1]
```diff
@@ -129,6 +121,15 @@ int ossl_rsa_padding_add_PKCS1_OAEP_mgf1_ex(unsigned char *to, int tlen,
...
 
+int RSA_padding_add_PKCS1_OAEP(unsigned char *to, int tlen,
+                               const unsigned char *from, int flen,
+                               const unsigned char *param, int plen)
+{
+    return ossl_rsa_padding_add_PKCS1_OAEP_mgf1_ex(to, tlen, from, flen,
+                                                   param, plen, NULL, NULL);
+}
+
+
```
Comment (PlaidCat) [inline crypto/rsa/rsa_oaep.c:132]: Why was this moved?

File: crypto/fips/fips_kdf_selftest.c [hunk 2]
```diff
@@ -361,7 +364,9 @@ static int FIPS_selftest_krb5kdf(void)
     EVP_KDF_CTX_free(kctx);
     return ret;
 }
+#endif
 
+#if 0
```
Comment (jason-rodri) [inline crypto/fips/fips_kdf_selftest.c:369]: @jallisonciq this is single step KDF or one-step this should be approved

File: crypto/aria/aria.c [hunk 3]
```diff
@@ -1110,6 +1133,12 @@ void aria_encrypt(const unsigned char *in, unsigned char *out,
                   const ARIA_KEY *key)
 {
     assert(in != NULL && out != NULL && key != NULL);
+
+    if (FIPS_mode()) {
+        FIPSerr(ERR_LIB_FIPS, FIPS_R_NON_FIPS_METHOD);
+        return;
```
Comment (PlaidCat) [inline crypto/aria/aria.c:1139]: is the the right thing to do here?

They seem to be asserting they are null ... but if we bailed early the data in `*out` will be unmodified and `do_encrypt` is also a void.

Should we be asserting instead and treat it like something is `NULL` in the input parameters?

File: crypto/asn1/standard_methods.h [hunk 4]
```diff
@@ -59,3 +59,61 @@ static const EVP_PKEY_ASN1_METHOD *standard_methods[] = {
 #endif
 };
 
+static const EVP_PKEY_ASN1_METHOD *standard_methods_fips[] = {
```
Comment (PlaidCat) [inline crypto/asn1/standard_methods.h:62]: These are all the supported FIPS-140-3 methods or just the ones we're actually going to submit?

File: crypto/evp/e_aes.c [hunk 5]
```diff
@@ -4277,10 +4302,10 @@ static int aes_ocb_cleanup(EVP_CIPHER_CTX *c)
     return 1;
 }
 
-BLOCK_CIPHER_custom(NID_aes, 128, 16, 12, ocb, OCB,
```
Comment (jason-rodri) [inline crypto/evp/e_aes.c:4280]: do we not need the 'BLOCK_CIPHER_custom' function any more?

File: crypto/rsa/rsa_oaep.c (hunk 1 above)
Comment (jallisonciq) [reply crypto/rsa/rsa_oaep.c:132]: To remove a warning of missing function definition. Previously this function was used before definition. Moving it to before it is used removes that warning.

File: crypto/fips/fips_kdf_selftest.c (hunk 2 above)
Comment (jallisonciq) [reply crypto/fips/fips_kdf_selftest.c:369]: These aren't removed for algorithm completeness, but to remove an "unused but defined function" warning. These are static functions that are not referenced *anywhere* else in this file. So they can be commented out to remove the warning.

File: crypto/aria/aria.c (hunk 3 above)
Comment (jallisonciq) [reply crypto/aria/aria.c:1139]: Yes it is the correct thing to do for FIPS. Any time a non-approved function is called, we need to log that fact (that an attempt was made to use a non-approved function). So it's not the same as the NULL input parameter case.

File: crypto/asn1/standard_methods.h (hunk 4 above)
Comment (jallisonciq) [reply crypto/asn1/standard_methods.h:62]: These are the ones we support - matches our published security policy.

File: crypto/aria/aria.c (hunk 3 above)
Comment (jason-rodri) [reply crypto/aria/aria.c:1139]: asserting did cross my mind, the only other place we asserted is when the program tried to exit FIPS mode.

File: crypto/evp/e_aes.c (hunk 5 above)
Comment (jallisonciq) [reply crypto/evp/e_aes.c:4280]: Yes we still do. It's used for AES in gcm mode, which we support. Here we are switching  to BLOCK_CIPHER_custom_fips to prevent AES from being used in ocb mode, which we are disallowing under FIPS.

Note the "ocb" parameter to this function - that's telling the macro expansion what cipher mode we're defining here.

File: crypto/aria/aria.c (hunk 3 above)
Comment (jallisonciq) [reply crypto/aria/aria.c:1139]: Won't asserting terminate the program ? That would break openssl speed, which tries to instantiate (and check for NULL return) to determine if we need to test the speed of a particular algo.

Calling FIPSerr in these cases are a failure indication that can be read by the calling program.

File: crypto/rsa/rsa_oaep.c (hunk 1 above)
Comment (jallisonciq) [reply crypto/rsa/rsa_oaep.c:132]: I *hate* warnings in code :-). They hide other errors (and indeed whilst I was tidying up warnings I found some genuine bugs that I fixed).

**Review by jason-rodri — APPROVED** [review]

**Comment by jallisonciq** [comment]
After comments from Maple and Solar on some of the design decisions here, I'm closing out this MR and we push a new one with the requested changes.
